### PR TITLE
Add LinesTracker, improve ImageProcessor and fix PointsTracker

### LINKED ISSUE #11: Implement lines tracking using OpenCV LineSegmentDetector
I had already implemented reference code in Matlab. I need to find it and make it into proper classes.

### PR DESCRIPTION
**Pull request by prashanthr05**
This PR introduces multiple changes,
- Forces `clang-format` in files in `src/Perception/src/PinHoleCamera.cpp` and `src/Perception/tests/Features/ArucoDetectorExample.cpp`
- Fixes bug in `PointsTracker` class and improves the `PointsTrackerTest` (See https://github.com/dic-iit/kindyn-vio/issues/10#issuecomment-888268130 for results).
- Adds `LinesTracker` class and `LinesTrackerTest` and relevant modifications to the composing class `ImageProcessor`. (See https://github.com/dic-iit/kindyn-vio/issues/11#issuecomment-889095082 for results). The LinesTracker class uses OpenCV's FastLineDetector for detecting lines and KLT tracker for tracking.

cc @traversaro @S-Dafarra let me know if it's ok to merge into master directly or if you would be interested in a quick review?

Closes #11.

### CODE REVIEW

**Comment by traversaro** [comment]
> cc @traversaro @S-Dafarra let me know if it's ok to merge into master directly or if you would be interested in a quick review?

Feel free to merge directly for me. I can look at the PRs even after it has been merged.

**Comment by S-Dafarra** [comment]
> cc @traversaro @S-Dafarra let me know if it's ok to merge into master directly or if you would be interested in a quick review?

I think it is up to you. Which kind of review would you prefer? :grin:

**Comment by prashanthr05** [comment]
> > cc @traversaro @S-Dafarra let me know if it's ok to merge into master directly or if you would be interested in a quick review?
> 
> I think it is up to you. Which kind of review would you prefer?

A very high-level review on readability of the code would be sufficient, at this stage.

File: src/Perception/src/ImageProcessor.cpp [hunk 1]
```diff
@@ -211,16 +252,84 @@ bool ImageProcessor::getImageWithDetectedFeatures(cv::Mat& outImg)
...
+    return true;
+}
+
+void ImageProcessor::Impl::drawPoints(const cv::Mat& img,
+                                      const TrackedPoints& points,
+                                      const bool& printMetaData)
+{
+    const int windowSize{10};
+    const int radius{2};
+    const int thickness{2};
```
[the remark this answers is not in the document]
Comment (prashanthr05) [reply src/Perception/src/ImageProcessor.cpp:273]: Yes, at the moment, these are hard-coded. Also all of the parameters in the `PointsTracker` class and `LinesTracker` class are intentionally left hard-coded. I will move them within an `initialize()` method in future iterations.

I will add a comment accordingly,thanks!

File: src/Perception/src/PinHoleCamera.cpp [hunk 2]
```diff
@@ -131,18 +131,16 @@ void PinHoleCamera::unprojectPoints(const std::vector<cv::Point2f>& uvs,
...
+void PinHoleCamera::unprojectPoint(const cv::Point2f& uv, cv::Point2f& xc)
 {
     const auto& fx{m_K(0, 0)};
     const auto& fy{m_K(1, 1)};
     const auto& cx{m_K(0, 2)};
     const auto& cy{m_K(1, 2)};
-    xc.x = (uv.x - cx)/fx;
-    xc.y = (uv.y - cy)/fy;
+    xc.x = (uv.x - cx) / fx;
+    xc.y = (uv.y - cy) / fy;
```
[the remark this answers is not in the document]
Comment (prashanthr05) [reply src/Perception/src/PinHoleCamera.cpp:141]: They're are floats/doubles. I am adding a check for `fx` and `fy` to be non-null in the initialize method of this class. Thanks!

File: src/Perception/src/ImageProcessor.cpp [hunk 3]
```diff
@@ -30,6 +31,10 @@ class ImageProcessor::Impl
 {
 public:
     bool trackPoints();
+    bool trackLines();
+
+    void drawPoints(const cv::Mat& img, const TrackedPoints& points, const bool& printMetaData);
+    void drawLines(const cv::Mat& img, const TrackedLines& trackedLines, const bool& printMetaData);
```
[the remark this answers is not in the document]
Comment (prashanthr05) [reply src/Perception/src/ImageProcessor.cpp:37]: Actually, there's a parameter to configure the verbosity in the `initialize` method which affects the `debug` member.
I wanted to have these methods to be independent of the data members, which is why these are booleans been used as parameters for these functions.

File: src/Perception/src/PinHoleCamera.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (prashanthr05) [reply src/Perception/src/PinHoleCamera.cpp:149]: Done in https://github.com/dic-iit/kindyn-vio/pull/23/commits/41e86c9bff4fc105e019c7e51fc94c5313afc102.

File: src/Perception/src/ImageProcessor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (prashanthr05) [reply src/Perception/src/ImageProcessor.cpp]: I changed these to be configured from parameters already. https://github.com/dic-iit/kindyn-vio/pull/23/commits/41e86c9bff4fc105e019c7e51fc94c5313afc102 (one less TODO to do.)

File: src/Perception/src/ImageProcessor.cpp [hunk 4]
```diff
@@ -211,16 +252,84 @@ bool ImageProcessor::getImageWithDetectedFeatures(cv::Mat& outImg)
...
+    }
+}
+
+void ImageProcessor::Impl::drawLines(const cv::Mat& img,
+                                     const TrackedLines& trackedLines,
+                                     const bool& printMetaData)
+{
+    const auto& nrLines = trackedLines.lines.size();
+    const int thickness{2};
+    const int windowSize{10};
```
[the remark this answers is not in the document]
Comment (prashanthr05) [reply src/Perception/src/ImageProcessor.cpp]: Done in https://github.com/dic-iit/kindyn-vio/pull/23/commits/41e86c9bff4fc105e019c7e51fc94c5313afc102

File: src/Perception/include/KinDynVIO/Perception/Features/LinesTracker.h [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    long long int m_featureID{0};
+
+    cv::Mat m_mask;
+    std::vector<cv::Vec4f> m_detectedLines;
+    std::vector<Line> m_forwardedLines, m_prevLines, m_newLines;
+
+    int m_trackingPixelErrThreshold{3};
+
+    int m_minDistanceBetweenFeatures{30}; // choose this size such that mask applied does not introduce spuriously detected lines, preferably above 15
+    int m_borderSize{1}; // image border size to check for interior tracked features
```
[the remark this answers is not in the document]
Comment (prashanthr05) [reply src/Perception/include/KinDynVIO/Perception/Features/LinesTracker.h]: Fixed in https://github.com/dic-iit/kindyn-vio/pull/23/commits/41e86c9bff4fc105e019c7e51fc94c5313afc102

File: src/Perception/include/KinDynVIO/Perception/Features/LinesTracker.h [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    cv::Point2f startPoint;      // normalized unprojected
+    cv::Point2f endPoint;        // normalized unprojected
+};
+
+struct TrackedLines
+{
+    std::vector<Line> lines;
+    std::vector<long long int> ids;
+    std::vector<long long int> counts;
+};
```
[the remark this answers is not in the document]
Comment (prashanthr05) [reply src/Perception/include/KinDynVIO/Perception/Features/LinesTracker.h]: At the moment, these structs were meant to be placeholders for items that are to be manipulated as a whole. I will keep them as it is for now and improve them in future iterations with the necessary safety handling.

Renamed them in https://github.com/dic-iit/kindyn-vio/pull/23/commits/41e86c9bff4fc105e019c7e51fc94c5313afc102.

**Comment by prashanthr05** [comment]
Ubuntu CI failure due to GTSAM, unrelated to the changes in this PR. macOS CI passed. I proceed with merging.

**Pull request merged**